Add support for multiple Valets within the same access group

efirestone opened the pull request:
Previously, the shared access group identifier was used for both the access group itself, as well as the uniqueness identifier for the given Valet. This adds an optional additional `identifier` parameter that can be specified when creating a shared access group Valet. The identifier adds an additional element of uniqueness, so two Valets with the same shared access group can exist, and their data will not overlap, so long as they have different identifiers.
    
The default for this value is `nil`, which keeps the existing behavior for full backward compatibility.

dfed commented on Gemfile:
```diff
@@ -1,3 +1,5 @@
+source "https://rubygems.org"
```
can we remove the `source 'https://rubygems.org' do` below? Seems duplicative to have both

dfed replied:
yeah we should be able to remove the `do/end` block below. [Example](https://github.com/dfed/swift-async-queue/blob/a232ddbfb776a3d3e380dce30e7d5f617cffcf82/Gemfile#L3-L5).

dfed commented on Sources/Valet/Valet.swift:
```diff
@@ -40,19 +40,21 @@ public final class Valet: NSObject {
-    ///   - identifier: The identifier for the Valet's shared access group. Must correspond with the value for keychain-access-groups in your Entitlements file.
+    ///   - groupIdentifier: The identifier for the Valet's shared access group. Must correspond with the value for keychain-access-groups in your Entitlements file.
     ///   - accessibility: The desired accessibility for the Valet.
+    ///   - identifier: An optional additional uniqueness identifier. Using this identifier allows for the creation of multiple Valets within the same shared access group.
```
possible clarity improvement:
```suggestion
    ///   - identifier: An optional additional uniqueness identifier. Using this identifier allows for the creation of separate, sandboxed Valets within the same shared access group.
```

dfed commented on Tests/ValetIntegrationTests/ValetIntegrationTests.swift:
```diff
@@ -317,6 +318,19 @@ class ValetIntegrationTests: XCTestCase
         }
     }
 
+    func test_stringForKey_withDifferingIdentifierInSameAccessGroup_throwsItemNotFound() throws
```
🎉

efirestone replied:
Done!

dfed commented on Tests/ValetIntegrationTests/ValetIntegrationTests.swift:
```diff
@@ -650,6 +677,22 @@ class ValetIntegrationTests: XCTestCase
+            XCTAssertEqual(error as? KeychainError, .itemNotFound)
+        }
+    }
+
```
It might be nice to add a test that migrating between two different valets with different identifiers works as expected. (I know the test matrix is big + annoying but the more tests we have the more confident we are that this stuff actually works)

efirestone replied:
The migration matrix doesn't seem to care about the *types* of the Valets that are involved in the migration, and more about the variations and keys and values. For example, there's nothing testing the migration of iCloud Valets to group access Valets to local Valets, etc. And I think that makes sense given that the migration uses the standard get/set/remove methods in all cases, and those are tested separately for each type.

Happy to add something here, but not entirely sure which tests of value are missing.

dfed replied:
Eh that's fair. I was going off memory. If we don't have similar tests then we're good to go here. I would, however, like to see test coverage improve (based on the code of comment), but I'm currently in mobile so I haven't checked details. Let me know if I can help offer possible tests to up coverage.

efirestone replied:
The Codecov report says `The diff coverage is 100.00%.`, so I think all the new content is covered. It went down overall, but I'm not sure how that's calculated.

dfed reviewed (commented):
Overall this seems reasonable to me. We're taking on a bit more complexity, but it's not too shabby. I'd love to better understand the use case for this, and why we're only adding this functionality to shared access Valets: is the use case specific to shared access Valets or is your current need currently scoped to shared access Valets?

Also: thoughts on updating the README to document this new API?

efirestone commented:
> Is the use case specific to shared access Valets or is your current need currently scoped to shared access Valets?
Our use case is only for shared group access Valets. The same thing could easily be extended to iCloud and other shared Valets as well, but I figured I'd start here to scope it down. Happy to add the other cases in this PR or a follow up too though.
 
> Also: thoughts on updating the README to document this new API?
Yup, good call. On it.

efirestone commented:
> > Is the use case specific to shared access Valets or is your current need currently scoped to shared access Valets?
> 
> Our use case is only for shared group access Valets. The same thing could easily be extended to iCloud and other shared Valets as well, but I figured I'd start here to scope it down. Happy to add the other cases in this PR or a follow up too though.
> 
> > Also: thoughts on updating the README to document this new API?
> 
> Yup, good call. On it.

Updated the README and added the ability to add uniqueness identifiers to the other shared types (iCloud, SecureEnclave, SinglePromptSecureEnclave).

dfed commented on README.md:
```diff
@@ -181,6 +190,8 @@ VALValet *const myCloudValet = [VALValet iCloudValetWithIdentifier:@"Druidia" ac
 
 This instance can be used to store and retrieve data that can be retrieved by this app on other devices logged into the same iCloud account with iCloud Keychain enabled. If iCloud Keychain is not enabled on this device, secrets can still be read and written, but will not sync to other devices. Note that  `myCloudValet` can not read or modify values in either `myValet` or `mySharedValet` because ` ...
 
+Similar to with shared group Valets, shared iCloud Valets can be created with an additional identifier, allowing multiple independently sandboxed keychains to exist within the same iCloud shared group.
```
```suggestion
As with Valets, shared iCloud Valets can be created with an additional identifier, allowing multiple independently sandboxed keychains to exist within the same iCloud shared group.
```

dfed commented on README.md:
````diff
@@ -169,6 +169,15 @@ VALValet *const mySharedValet = [VALValet sharedGroupValetWithGroupPrefix:@"grou
 
 This instance can be used to store and retrieve data securely across any app written by the same developer that has `group.Druidia` set as a value for the `com.apple.security.application-groups` key in the app’s `Entitlements`. This Valet is accessible when the device is unlocked. Note that `myValet` and `mySharedValet` cannot read or modify one another’s values because the two Valets were create ...
 
+It is possible to create multiple Valets within the same shared access group be specifying the same group identifier but different unique identifiers, like so:
+
+```
+let valet1 = Valet.sharedGroupValet(with: groupID, identifier: Identifier(nonEmpty: "valet1")!, accessibility: accessibility)
+let valet2 = Valet.sharedGroupValet(with: groupID, identifier: Identifier(nonEmpty: "valet2")!, accessibility: accessibility)
+```
+
+These two Valets cannot access each others data, but they are both accessible from all applications or extensions which have access to the given `groupID`. The `identifier` (but not the `groupIdentifier`) is optional on shared Valets. All Valets created within the same group using the default `identifier` can access the same data.
+
````
let's move this blurb up to the `Valet` section and then replace this with:
```suggestion
As with Valets, shared iCloud Valets can be created with an additional identifier, allowing multiple independently sandboxed keychains to exist within the same shared group.

```

dfed replied:
@efirestone we still want this blurb but we want it further up in the README 😉

efirestone replied:
I'm not sure I'm following what you're suggesting (and now the review recommendation can't be committed). Are you saying put this blurb in the `Valet` section:
```
As with Valets, shared iCloud Valets can be created with an additional identifier, allowing multiple independently sandboxed keychains to exist within the same shared group.
```
or to leave that blurb in the "Sharing Secrets" section, and then put the original text up in the Valet section?

dfed replied:
Ah! Okay I see where I failed to read here. It's the _additional_ in `additional identifier` that threw me.

The _additional_ is referring to iCloud Valets, whereas Valets have a primary identifier.

I think we should just remove the "As with Valets" preamble and this'll make more sense to me.

dfed reviewed (approved):
LGTM with a few changes to the README.

efirestone commented:
@dfed can you merge this? I don't have permissions 🙏

efirestone commented:
"As with Valets" prefix removed in the README, so should be good to go.

The iOS 13 + iOS 14 builds aren't finishing anymore, I'm guessing due to a GitHub change. It looks like `master` is having the [same problem now](https://github.com/square/Valet/pull/299). I'm investigating some in that PR, but so far I'm kind of stuck.

dfed commented:
> "As with Valets" prefix removed in the README, so should be good to go.
> 
> 
> 
> The iOS 13 + iOS 14 builds aren't finishing anymore, I'm guessing due to a GitHub change. It looks like `master` is having the [same problem now](https://github.com/square/Valet/pull/299). I'm investigating some in that PR, but so far I'm kind of stuck.

Thanks for addressing!

Re Xcode 12.4: can you file an issue on https://github.com/actions/runner-images? I've had success there in the past when there was a runner configuration issue.

dfed commented:
Just filed an issue for what we're seeing with CI: https://github.com/actions/runner-images/issues/7687

dfed commented:
Just merged latest into this branch. Should turn green, at which point merge at will @efirestone. Let me know if you want help publishing the feature version update.

dfed commented:
Just released this as 4.2.0
